src/net_processing.cpp

```diff
@@ -3198,6 +3237,117 @@ void PeerManagerImpl::ProcessValidTx(NodeId nodeid, const CTransactionRef& tx, c
     }
 }
 
+void PeerManagerImpl::ProcessPackageResult(const Package& package, const PackageMempoolAcceptResult& package_result, const std::vector<NodeId>& senders)
+{
+    AssertLockNotHeld(m_peer_mutex);
+    AssertLockHeld(g_msgproc_mutex);
+    AssertLockHeld(cs_main);
+
+    if (package_result.m_state.IsInvalid()) {
+        m_recent_rejects_reconsiderable.insert(GetPackageHash(package));
+    }
+    // We currently only expect to process 1-parent-1-child packages. Remove if this changes.
+    if (!Assume(package.size() == 2)) return;
+
+    // No package results to look through for PCKG_POLICY or PCKG_MEMPOOL_ERROR
+    if (package_result.m_state.GetResult() == PackageValidationResult::PCKG_POLICY ||
+        package_result.m_state.GetResult() == PackageValidationResult::PCKG_MEMPOOL_ERROR) return;
```

Comment:
I wonder if there are situations in which we fail on a package level and therefore return here and never call `ProcessInvalidTx()` to remove the child from the orphanage. Would that be possible, for example, with v3 transactions that violate `PackageV3Checks()`, and is there something we could do against it? (or should, maybe it's not that terrible?).

Reply:
`PCKG_POLICY` doesn't populate `m_tx_results` so we wouldn't have results for those transactions - it's meant for situations in which the package is invalid, but the individual transactions could be if submitted in a different context. However, going over the failures that are marked as `PCKG_POLICY`, I agree there's a few failures that are classified as `PCKG_POLICY` but are actually more definitively invalid, and we should bubble that up to p2p.

Having both mempool + package v3 parents, v2-spending-v3, and v3-spending-v2 are failures that can't change in another context, so we can delete the tx from orphanage.

We shouldn't do this for `package-mempool-limits` and `package-not-child-with-unconfirmed-parents`, since there could be a parent with a smaller witness that makes this child acceptable.

The 2 changes we'd make are:
- update validation-side things to return a `MempoolAcceptResult` for the results where we should mark transactions as invalid. Perhaps we can reclassify these failures as `PCKG_TX` failures, or allow `PCKG_POLICY` failures to populate `m_tx_results`. cc @instagibbs ?
- update the p2p logic here to continue when it's `PCKG_POLICY` and call `ProcessInvalidTx` for the `MempoolAcceptResult`s we find.

Reply:
> PCKG_POLICY doesn't populate m_tx_results so we wouldn't have results for those transactions

From my reading of the code, `package-mempool-limits` populates `m_tx_results`? We also allow those cases in our fuzz target(which I'm guessing is there because I hit it!).

We could probably be more aggressive about filling out results when possible and passing those results along, sounds like future work for now?

Reply:
Then maybe we should just do this?

> Update the p2p logic here to continue when it's PCKG_POLICY and call ProcessInvalidTx for the MempoolAcceptResults we find.

(and remove the `Assume` that they always exist)

Reply:
^I added this to #30012